File: 210-course-schedule-ii/210. Course Schedule II.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        preq = defaultdict(list)
        for u,v in prerequisites:
            preq[u].append(v)

        self.res = []
        visited = set()
        cycle = set()
        def dfs(course):
            if course in cycle:
                return False
            if course in visited:
                return True

            cycle.add(course)

            for p in preq[course]:
                if not dfs(p):
                    return False

            visited.add(course)
            cycle.remove(course)
            self.res.append(course)
            return True
        
        for i in range(numCourses):
            if not dfs(i):
                return []
        return self.res

File: 210-course-schedule-ii/210. Course Schedule II.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int, int> indegree;
        unordered_map<int, vector<int>> graph;
        vector<int> ordering;

        for (auto& arr : prerequisites) {
            int course = arr[0], preq = arr[1];
            graph[preq].push_back(course);
            indegree[course]++;
        }
        cout << "reached" << endl;
        queue<int> q;
        unordered_set<int> taken_courses;
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                q.push(i);
                taken_courses.insert(i);
                ordering.push_back(i);
            }
        }

        while (!q.empty()) {
            int course = q.front();
            q.pop();
            for (auto& nei : graph[course]) {
                indegree[nei]--;
                if (indegree[nei] == 0) {
                    q.push(nei);
                    taken_courses.insert(nei);
                    ordering.push_back(nei);
                }
            }
        }

        return taken_courses.size() == numCourses ? ordering : vector<int>{};
    }
};